prototype of tbb executor

jspark1105 opened the pull request:
Summary:
Explore using TBB for both inter-op and intra-op parallelization. This diff also includes a simple MLP benchmark that uses multiple numa nodes to compare its performance with OpenMP.

This diff is for prototyping and to get help/feedback from Intel TBB team.
This is not intended for landing (at least in the current form).

Differential Revision: D8556648

mingfeima commented:
@jspark1105 how does this go so far? are you trying to compare `tbb` + `single thread mkl` v.s. openmp here?

jspark1105 commented:
@mingfeima we're more interested in using tbb for intra-op and inter-op parallelism (intra-op parallelism done with tbb + single thread MKL) compared with using MKL for intra-op and Caffe2 thread pool for inter-op. Mike, Alexey, and Alexei in TBB team have been providing great help to us. I haven't been able to make progress in the last few weeks but should be able to look at this again this week.

Commit:
TBB task graph

Summary: Adding an alternative implementation of a task graph based on TBB

fbshipit-source-id: cb3cb37445cb97271205f88d39dd118a8b6420eb

Commit:
Task graph and task future abstractions in executor

Summary:
Based on ideas from JIT-based executor prototype, this diff refactors executor
and introduces abstractions of task graph, task and task future.

Right now executor occupies the space between lower-level (ops and their
hardware events) and higher-level layers (e.g. DPER). Executor is
responsible for multiple things:
- interfacing with higher-level code, reading execution flags, net arguments;
- actually running ops on different thread pools;
- keeping track of dependencies between ops (chains), e.g. technicalities like
  keeping track of a number of already processed parent chains and so on;
- scheduling logic that links lower-level events with graph task
  scheduling (e.g. taking care of async CPU ops, taking care of CUDA->CPU case)

In this diff we further decompose executor into:
- Task graph - abstracts away actual implementation of task scheduling; we keep
  the current implementation (AsyncTaskGraph, based on futures) but also
  experiment with TBB-based one. For the purposes of current implementation, we
  use task and future abstractions
- Task (AsyncTask) - wraps chain of events, executes a sequence of RunAsync calls,
  manages a future (typically sets it to the completed state after running ops
  or, as in case of async cpu ops, links future to async cpu op's Event)
- Task future (AsyncTaskFuture) - future without value, supports callbacks
    - isolates low-level technicalities like keeping track of processed parents
      (combined future is created from multiple futures)

The main purpose of this diff is to decompose complicated logic of schedule()
function into multiple easy to understand pieces and allow alternative
implementations (e.g. based on TBB):
- Tasks run ops (run sequence of RunAsync calls) and are responsible for
  managing a single task's future
- Task graph creates tasks and setups dependencies between them through future
  callbacks
- When future is set to a completed state, we assume that we may proceed to
  scheduling a depending op (multiple dependencies are handled though a combined
  future). Note, for example, that we no longer need to consider the special
  case of async cpu op when setting up dependencies
- The only special case to handle - CUDA->CPU, handled by a future callback
- Note that this simplifies the scheduling logic, e.g.
  - no need for special logic to keep track of number of processed tasks and
    when to call finishRun
  - no need to have a separate running_cv_ condition variable
  - replacing multiple parentCallback callbacks for a pending task with a
    single combined future and a callback, etc
  - all the above is implemented using futures and callbacks

Further, we reduce the usage of current C2 Event to (in case of
CUDA ops, for now) a wrapper around CUDA event and (in case of async CPU ops) an
interface between async cpu op and executor (SetFinished/SetFinishedWithException).

Differential Revision: D13043261

fbshipit-source-id: 55c1bba374db3abdc48f3230031a93a08167cdda

jspark1105 committed:
TBB FC and FCGradient ops (#9460)

Summary:
Moved net_tbb to operators/tbb from core

Pull Request resolved: https://github.com/pytorch/pytorch/pull/9460

Explore using TBB for both inter-op and intra-op parallelization. This diff also includes a simple MLP benchmark that uses multiple numa nodes to compare its performance with OpenMP.

This diff is for prototyping and to get help/feedback from Intel TBB team.
This is not intended for landing (at least in the current form).

Differential Revision: D8556648

fbshipit-source-id: 86b304ec12e09c08436bb77baecaa3c23cdcbbdc
